### ISSUE TITLE
Add ability to control feature altitude via a style spec property

### ISSUE THREAD

**Issue by lucaswoj** [mapbox/mapbox-gl-js#3993]
Continued from https://github.com/mapbox/mapbox-gl/issues/6

> @peterqliu 
> We should take advantage of the extra dimension that map pitching affords us, and use depth to visualize overlapping elements that would otherwise obstruct each other.
> 
> ![screen shot 2015-12-15 at 4 51 20 pm](https://cloud.githubusercontent.com/assets/4551508/11907508/534aaa06-a589-11e5-92ab-c68e9e42f599.png)
> 
> Currently, the layer stack order is dictated by the sequence it was specified in the stylesheet; when applied, the proposed z-offset would override that. This is analogous to HTML's natural stacking order overridden by `z-index` in CSS.
> 
> Also, label rendering order is currently tightly bound to layers' z-order, because they both key off of the stylesheet layer sequence. Adding z-offset will free us to decide which layers get rendering priority, regardless of whether they go above or below other layers.
> 
> cc @lucaswoj @kkaefer 

cc @tmcw @samanpwbb @jfirebaugh @ansis

**Comment by ashleyclough** [comment]
Allowing polygons and symbols to have an altitude or z value would then mean that data presented by fill-extrusion could be represented and interrogated consistent with the 3D presentation [mockup]:

![3d polygon and symbol](https://cloud.githubusercontent.com/assets/912721/22057822/50efb608-dd5e-11e6-94c9-40541f72cd09.jpg)

**Comment by robinbezak** [comment]
@ashleyclough  Ohh gooosh how u did the markers extrude?

**Comment by ashleyclough** [comment]
I didn't, the image is just a mockup, created in Photoshop, of how I would like it to work!

**Comment by patodevilla** [comment]
Is there any progress on this feature? I'm building a indoor 3D application and need to have markers or popups to offset from the terrain when the map is pitched. Is there an alternative solution to make this work?

**Comment by ansis** [comment]
We're currently working on a new approach to symbol placement (collision avoidance) that will be able to handle symbols with z offsets. After we finish that we can take a look at adding this to the spec and implementing it.

**Comment by mariokam** [comment]
Hi ansis, any updates on this topic? Is it in the roadmap?

**Comment by peterqliu** [comment]
@mariokam check out the roadmap [here](https://www.mapbox.com/mapbox-gl-js/roadmap/)

**Comment by mariokam** [comment]
Thanks peterqliu, and I know about this roadmap, but I would expect that big company/startup as Mapbox to have more detailed roadmap, than 4 features in progress and 4 feature in long-term. So I don't see any plans even for **z offsets**

**Comment by xkisu** [comment]
Any updates on this? I have need of this for a client project where we have several raster layers stacked that need an offset so you can tell them apart.

**Comment by ansis** [comment]
@xkisu, @mariokam the new approach to symbol placement is being ported from here to mapbox/mapbox-gl-native. We'll be able to start looking at this after that is done.

**Comment by tianbaolin** [comment]
I hope mapbox can support altitude in geojson or can set  height attribute in a layer..

**Comment by 1tylermitchell** [comment]
We're setting a [raster layer height value](https://github.com/xkisu/mapbox-gl-js/commit/faebae0d3eee147a98fd1060cb564a348c43e353) and it does draw the layer but only at certain angles.  Will this issue (above) address that?  It looks like some sort of clipping occurs depending on angle.  Any tip?
<img width="1199" alt="mapbox_clip_error" src="https://user-images.githubusercontent.com/518729/33058007-58d0634c-ce42-11e7-9510-018186e8fc17.png">

**Comment by shongololo** [comment]
Another vote for supporting relatively basic 3d features in the sense of being able to show overlapping features in 3d space (map pitch) - i.e. a bridge crossing a road, or multiple floor levels inside a building shell. i.e. very simple support for a z-height coordinate, not unlike PostGIS supports a height coordinate.

**Comment by saferthanhouses** [comment]
This is something that we'd very much like to see and seems like a necessity in a platform that supports 3d mapping. Are there any updates on a timeline for this? Is this functionally absent in the native-sdks also?

**Comment by janschmutz** [comment]
Any updates on this?

**Comment by leehoss** [comment]
Any updates on this???

**Comment by fea17e86** [comment]
We need a feature, like being able to set the z offset or base height to speak in fill extrusion lang, too. Please give us feedback about related plans.

**Comment by maroonedione** [comment]
Any updates on this????

**Comment by asheemmamoowala** [comment]
@maroonedione  @fea17e86 @leehoss This is on our radar, but not currently on our roadmap over the next 6 months.

**Comment by cl-nati** [comment]
We on ClimaCell want it too.
Upvote! ✌️

**Comment by animeshjn** [comment]
+1
We really want this feature of Z-offsets per point of a line

**Comment by likeswinds** [comment]
+1

**Comment by 1Map** [comment]
+1

**Comment by mayteio** [comment]
+1 - need to label extruded buildings at the top.

**Comment by narrowtux** [comment]
stop getting my hopes up with the pointless `+1` bumps :P

**Comment by rajputvishal** [comment]
https://deck.gl/#/examples/overview?section=core-layers is able to do it and it can also be integrated with mapbox.

**Comment by asheemmamoowala** [comment]
Some work on this is being done in the [Vector Tile 3 specification](https://github.com/mapbox/vector-tile-spec/tree/v3.0-development/3.0#43-geometry-encoding). This is part of the team's long term roadmap, but there is not a timeframe for it as yet.

**Comment by azhdarshirinzada** [comment]
any updates?

**Comment by arcsump** [comment]
How come wrld.js is able to apply this using the elevation property in marker?

**Comment by andrewharvey** [comment]
> How come wrld.js is able to apply this using the elevation property in marker?

This ticket is specifically about layers in the style spec, so excludes Marker elements, I think it's worth opening a separate ticket for Marker elevation/z-offset.

**Comment by patrickkeenan** [comment]
> This ticket is specifically about layers in the style spec, so excludes Marker elements, I think it's worth opening a separate ticket for Marker elevation/z-offset.

But this issue says discussion should be concentrated here no? 
https://github.com/mapbox/mapbox-gl-js/issues/8988

**Comment by andrewharvey** [comment]
> But this issue says discussion should be concentrated here no?

Ok, Personally I see them as different issues still, one is for layers part of the style spec, the other is for Marker elements. Different implementation and different API.

**Comment by derrickrung** [comment]
Hoping to bring some attention back to this...

Any updates on this?

**Comment by cigone-openindoor** [comment]
three.js could help...

**Comment by arcsump** [comment]
I'll be coming up with a more detailed use-case for this, but basically , a way to set the point's elevation is vital for visualizing points at different levels.

Our motivation is, we are surveyors working for a major tunnel project here in AU , and part of what we visualize is the monitoring status of the points for critical supporting structure. We are building a web-based site to be able to monitor the status.

Currently, the **points** in the area we are monitoring at the moment, **look flat**, 
![image](https://user-images.githubusercontent.com/5513496/94511446-b614df00-025c-11eb-8456-94c1e1c8ae2d.png)

when at least we wanted to reflect the **multi-level nature of the points.**
![image](https://user-images.githubusercontent.com/5513496/94511643-3b988f00-025d-11eb-8e28-5b019224d1bf.png)


And, I am not yet referring to the **negative extrusion heights** that should be useful in visualizing the whole area - this is part of an **underground** tunnel infrastructure  ( which I would **happily ask a feature request for separately.**)

If someone can refer me to the **format I can use to present this use-case, it would be helpful.**

I hope we can push for the development of this feature, even just the support for the elevation for point layers.

**Comment by mayteio** [comment]
@arcsump Metro tunnel? I work for CoM across the road! Have you tried any esri arcgis JS products? The developer experience is horrible but it’s pretty powerful when it comes to 3D, supporting both elevation and negative elevation.

**Comment by ats1999** [comment]
> well, i am just hit this when i want to show an indoor map.
> all is ok, except for the symbol on top of Extrude layer
> [my example url](https://api.mapbox.com/styles/v1/dbskcnc/cjf4s1tvg1lds2ro36557hfa3.html?fresh=true&title=true&access_token=pk.eyJ1IjoiZGJza2NuYyIsImEiOiJjaXUwZnpsbjIwMTVzMnpwZ3Z2bWl2OTFkIn0.-0yM74vhUs345reV9vKqTA#20.26/25.0036464/118.6217658/93.6/60)

Can i get the source code?

**Comment by arcsump** [comment]
> @arcsump Metro tunnel? I work for CoM across the road! Have you tried any esri arcgis JS products? The developer experience is horrible but it’s pretty powerful when it comes to 3D, supporting both elevation and negative elevation.

Yes, thank you for the suggestion - will definitely look at this and wrld.js as I am not sure what is the status of the request, as they pride on 3d stuff, but seems to have missed out on this functionality.

**Comment by arcsump** [comment]
an alternative is to use [Maptalks](https://maptalks.org/) until the feature can be implemented.

**Comment by arcsump** [comment]
> three.js could help...

In fact a complex solution was implemented for popups in #9933

**Comment by likwidgames** [comment]
Any updates on this? This has been open since 2017 and seems like an important feature that competitors already have in place.

**Comment by arcsump** [comment]
Unfortunately, the focus is on releasing 3D / terrain mapping instead.

You can use the excellent alternatives listed prior, like
ESRI ArcGIS for JS

or the threeBOX library which still integrates into the mapbox library.

Asked our devs at Soletanche Freysinnet France to look at alternatives for
the Vinci Group for now

**Comment by openSourcerer9000** [comment]
The people have spoken... we want more dimensionality!

**Comment by leehoss** [comment]
Yeah can we give this issue a bump up… this feature will change the game !

**Comment by cigone-openindoor** [comment]
I must admit that I do not use mapbox anymore, but maplibre.
So no more interest for me...

**Comment by spencerthayer** [comment]
Has there been any progress getting markers, points, or any object to render with a z-offset or altitude?

**Comment by arcsump** [comment]
If someone can actually explain **what's stopping Mapbox** from implementing the **3rd dimension** in this
[geojson spec](https://datatracker.ietf.org/doc/html/draft-butler-geojson#section-3), that would be helpful.

**Comment by JonPichel** [comment]
What's the current status of this? I have a Drone GCS which uses ArcGIS Maps SDK for Javascript but would love to switch to Mapbox, but without 3D I can't possibly do that. In my opinion the ability to draw markers, lines and models in 3D would suppose a huge grow for mapbox.

**Comment by leehoss** [comment]
Is there any update on this - do we really need to strip out markers and add them in 3js - seem like a straight forward concept - when we turn terrain on the labels align to the terrain - I'm sure there could be a property to set this value !! Please respond

**Comment by ashleyclough** [comment]
Have you seen the symbol-z-elevate property in the v3 style specifications? It won't do it automatically like the terrain alignment as you'll need to know each symbol's z height.

**Comment by jamesbarker185** [comment]
would love to be able to have 3d multi-lines on map

**Comment by astojilj** [comment]
Please check the Elevated Lines section in the latest blog post https://www.mapbox.com/blog/mapbox-style-updates-more-flexible-configurations-for-the-3d-basemap

Per-vertex elevation control is for now possible only with geoJSON source that has lineMetrics set, and using line-progress to access elevation from array data property, like it the render test here:

https://github.com/mapbox/mapbox-gl-js/blob/main/test/integration/render-tests/terrain/lines-elevated/style.json

geojson source:

```
"geojson1": {
      "type": "geojson",
      "lineMetrics": true,
      "data": {
        "type": "Feature",
        "properties": {
          "elevation": [400, 400, 400, 0, 400, 400, 400]
        },

```
  layer:

```
      "line-z-offset": [
          "at",
          ["floor", ["*", [ "line-progress" ], 6.9]],
          ["get", "elevation"]
        ]
```

_Stay tuned_ for additional elevation support and examples.

**Comment by Jxlle** [comment]
Any updates on this?

**Comment by Espade** [comment]
Any feature planned for this?

**Comment by astojilj** [comment]
@abernier ,
Please try the experimental _fill-z-offset_. An example of usage is in the render tests, e.g. 
https://github.com/mapbox/mapbox-gl-js/blob/main/test/integration/render-tests/fill-z-offset/property-function/style.json. cc @endanke 

@Jxlle @Espade 

We have integrated support for ground and sea level reference for elevated lines, and we're working on examples. cc @lasselammi

For sea level reference point for the flight path, please refer to the implementation in render tests meanwhile, e.g. https://github.com/mapbox/mapbox-gl-js/blob/main/test/integration/render-tests/terrain/lines-elevated-interpolated-elevation/style.json:

There, you can see smooth elevated lines over terrain. 

![image](https://github.com/user-attachments/assets/826f4194-5d4f-44cf-9480-463fffcd25d7)

Note that both features are are still under work - early feedback is more than welcome.

**Comment by Espade** [comment]
@astojilj great work and continue to @srmncnk question, do you plan this also for flutter release?

**Comment by endanke** [comment]
@srmncnk it's available on mobile as well!

@Espade with Flutter some of the experimental properties are not exposed via dedicated runtime APIs, however you can use them with the [setStyleLayerProperty](https://pub.dev/documentation/mapbox_maps_flutter/2.4.1/mapbox_maps_flutter/StyleManager/setStyleLayerProperty.html) function or by configuring it in the style JSON.

**Comment by jmegroot** [comment]
Do you have an example for implementing this?

**Comment by AkBo24** [comment]
Has anyone successfully used the `line-progress` modifier in swift? I can't seem to interpolate a slope with some mock data.


```
            GeoJSONSource(id: "SOURCE_ID")
                .data(.string(geoJsonFeature))
            
            // Specify a unique string as the layer ID (LAYER_ID)
            // and reference the source ID (SOURCE_ID) added above.
            LineLayer(id: "LAYER_ID", source: "SOURCE_ID")
                .lineJoin(.round)
                .lineZOffset(
                    Exp(.at) {
                        Exp(.product) {
                            Exp(.lineProgress)
                            12
                        }
                        Exp(.get, "elevation")
                    }
                )
                .lineCrossSlope(0)
                .lineElevationReference(.sea)
                .lineWidth(8)
                .lineColor(.red)
```

My inline GeoJSON:

```
let geoJsonFeature = """
{
  "type": "Feature",
  "properties": {
    "elevation": [10000, 15000, 20000, 25000, 30000, 35000, 40000, 45000, 50000, 55000, 60000, 65000]
  },
  "geometry": {
    "type": "LineString",
    "coordinates": [
      [21.79401104323395, 61.493343399275375],
      [21.6497937188623, 61.369485877583685],
      [21.723373986398713, 61.20121331932606],
      [21.579156662028026, 61.15722935505474],
      [21.517349237298163, 61.1174491345069],
      [21.532065290804866, 61.03916342463762],
      [21.611531979743546, 60.922086661997184],
      [21.82049993954618, 60.82754056454698],
      [22.02063826724438, 60.662131116075585],
      [22.129538402396946, 60.543665503992116],
      [22.206061880634536, 60.461061119997595],
      [22.270812516066485, 60.4538050749446]
    ]
  }
}
"""
```

**Comment by john-work-ios** [comment]
I'm still trying to figure out how to elevate a polyline on iOS, anyone had success?

**Comment by alienatorZ** [comment]
I am seeing an issue where the z-elevation does not work on a FeatureCollection of hundreds of features.

**Comment by NeluQi** [comment]
I'm really looking forward to this feature in the main version

I ideally need elevation for all map entities, including geojson and labels, features, and so on.
